sys/net/link_layer/ieee802154/submac.c

```diff
@@ -0,0 +1,384 @@
+        _tx_end(submac, TX_STATUS_NO_ACK, NULL);
+    }
+    else {
+        /* Perform retransmissions */
+        _perform_retrans(submac);
+    }
+}
+
+void ieee802154_submac_tx_done_cb(ieee802154_submac_t *submac)
+{
+    ieee802154_dev_t *dev = submac->dev;
+    ieee802154_tx_info_t info;
+
+    ieee802154_radio_confirm_transmit(dev, &info);
+
+    switch (info.status) {
+    case TX_STATUS_MEDIUM_BUSY:
+        _handle_tx_medium_busy(submac);
+        break;
+    case TX_STATUS_NO_ACK:
+        _handle_tx_no_ack(submac);
+        break;
+    case TX_STATUS_SUCCESS:
+    case TX_STATUS_FRAME_PENDING:
+        _handle_tx_success(submac, &info);
+        break;
+    default:
+        assert(false);
+        break;
+    }
+}
+
+int ieee802154_send(ieee802154_submac_t *submac, const iolist_t *iolist)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    uint8_t *buf = iolist->iol_base;
+    bool cnf = buf[0] & IEEE802154_FCF_ACK_REQ;
+
+    if (submac->state == IEEE802154_STATE_OFF) {
+        return -ENETDOWN;
+    }
+
+    if (submac->tx ||
+        ieee802154_radio_request_set_trx_state(dev,
+                                               IEEE802154_TRX_STATE_TX_ON) <
+        0) {
+        return -EBUSY;
+    }
+
+    submac->tx = true;
+
+    ieee802154_radio_write(dev, iolist);
+    while (ieee802154_radio_confirm_set_trx_state(dev) == -EAGAIN) {}
+
+    submac->wait_for_ack = cnf;
+    submac->retrans = 0;
+
+    ieee802154_csma_ca_transmit(submac);
+    return 0;
+}
+
+int ieee802154_submac_init(ieee802154_submac_t *submac)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    submac->tx = false;
+    submac->state = IEEE802154_STATE_LISTEN;
+
+    ieee802154_radio_request_on(dev);
+
+    /* generate EUI-64 and short address */
+    luid_get_eui64(&submac->ext_addr);
+    luid_get_short(&submac->short_addr);
+    submac->panid = CONFIG_IEEE802154_DEFAULT_PANID;
+
+    submac->be.min = CONFIG_IEEE802154_DEFAULT_CSMA_CA_MIN_BE;
+    submac->csma_retries = CONFIG_IEEE802154_DEFAULT_CSMA_CA_RETRIES;
+    submac->be.max = CONFIG_IEEE802154_DEFAULT_CSMA_CA_MAX_BE;
+
+    submac->tx_pow = CONFIG_IEEE802154_DEFAULT_TXPOWER;
+
+    if (ieee802154_radio_has_24_ghz(dev)) {
+        submac->channel_num = CONFIG_IEEE802154_DEFAULT_CHANNEL;
+        submac->channel_page = 0;
+    }
+    else {
+        submac->channel_num = CONFIG_IEEE802154_DEFAULT_SUBGHZ_CHANNEL;
+        submac->channel_page = CONFIG_IEEE802154_DEFAULT_SUBGHZ_PAGE;
+    }
+
+    /* If the radio is still not in TRX_OFF state, spin */
+    while (ieee802154_radio_confirm_on(dev) == -EAGAIN) {}
+
+    /* Enable Auto ACK */
+    ieee802154_radio_set_rx_mode(dev, IEEE802154_RX_AACK_ENABLED);
+
+    /* Configure address filter */
+    ieee802154_radio_set_hw_addr_filter(dev, &submac->short_addr,
+                                        &submac->ext_addr, &submac->panid);
+
+    /* Configure PHY settings (channel, TX power) */
+    ieee802154_phy_conf_t conf =
```

Comment:
```suggestion
    const ieee802154_phy_conf_t conf =
```

Reply:
I forgot to declare this as const in the radio HAL. I will open a PR to fix it